### PR TITLE
Add HTTP API

### PR DESCRIPTION
**Pull request by branden**
This PR adds a package that provides an HTTP API for checks, and an `http-server` CLI subcommand that exposes it. See `api/openapi.yaml` for an OpenAPI spec, and `README.md` for usage instructions.

https://jira.mesosphere.com/browse/DCOS_OSS-1406

DC/OS PR: https://github.com/dcos/dcos/pull/3037

### CODE REVIEW

**Review by klueska — COMMENTED** [review]
Did you mean to include a bunch of vendor changes in this commit 
df903d3? Because you have a seaprate commit for updating vendored changes later on.

File: cmd/http_server.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/dcos/dcos-check-runner/runner"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var httpServerCmd = &cobra.Command{
+	Use:   "http-server",
+	Short: "Start the check runner HTTP server",
+	Long:  `TODO`,
```
Comment (philipnrmn) [inline cmd/http_server.go]: one stray TODO

**Comment by philipnrmn** [comment]
Otherwise LGTM!

File: cmd/http_server.go (hunk 1 above)
Comment (branden) [reply cmd/http_server.go]: 😅

**Review by philipnrmn — DISMISSED** [review]

File: api/router.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	rh := runnerHandler{runner: runner}
+
+	base := router.PathPrefix(baseURI).Subrouter()
+	base.HandleFunc("/{check_type}/", rh.listChecks).Methods("GET")
+	base.HandleFunc("/{check_type}/", rh.runChecks).Methods("POST")
+
+	return router
+}
+
+type runnerHandler struct {
```
Comment (klueska) [inline api/router.go]: I would expect this to be above the `NewRouter()` definition

File: api/router.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+		http.Error(w, httpErr.Error(), httpErr.statusCode)
+		return
+	}
+
+	rs, err := checkFunc(r.Context(), true, checksFromQueryParams(r)...)
+	if err != nil {
+		http.Error(w, "Error listing checks", http.StatusInternalServerError)
+		return
+	}
+	writeJSONResponse(w, rs)
```
Comment (klueska) [inline api/router.go]: I would probably put a newline above this.

File: api/router.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+	return router
+}
+
+type runnerHandler struct {
+	runner *runner.Runner
+}
+
+func (rh *runnerHandler) listChecks(w http.ResponseWriter, r *http.Request) {
+	checkFunc, httpErr := rh.checkFuncFromReq(r)
```
Comment (klueska) [inline api/router.go]: I would name the function called here `getCheckFuncFromReq` or simply `getCheckFunc` (assuming there's not another `getCheckFuncFromXXX`) because the parameter type tells us it's from a request.

When I see a function called `checkFuncFromReq()`, this implies to me that you are actually going to "check" something inside the function, not return a "checkFunc" to me.

I do see the reasoning here though....
Functions in Go usually either:
1) begin with a verb, telling us what operation the function is going to perform; or
2) begin with a noun, implying a "getter" function for the noun listed

The problem here is that `check` can be interpreted as either a verb or a noun, leading to ambiguity. As such, I normally wouldn't recommend adding `get` to a function name since this is against normal best-practices for Go, but it actually seems appropriate here.

Can you think of an alternative name to avoid this confusion while still maintaining best-practices?

File: api/router.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		http.Error(w, "", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Write(body)
+}
+
+type httpError struct {
```
Comment (klueska) [inline api/router.go:207]: I would probably expect the code from here down to be at the top of this file.

File: api/router.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		http.Error(w, httpErr.Error(), httpErr.statusCode)
+		return
+	}
+
+	rs, err := checkFunc(r.Context(), false, checks...)
+	if err != nil {
+		http.Error(w, "Error running checks", http.StatusInternalServerError)
+		return
+	}
+	writeJSONResponse(w, rs)
```
Comment (klueska) [inline api/router.go]: Ditto about putting a newline above this

File: api/router.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func (rh *runnerHandler) listChecks(w http.ResponseWriter, r *http.Request) {
+	checkFunc, httpErr := rh.checkFuncFromReq(r)
+	if httpErr != nil {
+		http.Error(w, httpErr.Error(), httpErr.statusCode)
+		return
+	}
+
+	rs, err := checkFunc(r.Context(), true, checksFromQueryParams(r)...)
```
Comment (klueska) [inline api/router.go:50]: There's something strange here with me being able to intuitively understand what this function is going to do without digging into the guts of it.

Just from reading it, it looks like it is going to _actually_ perform some check based on the set of checks listed in the request's Query Parameters.

The meaning of the `bool` parameter is unclear.

Comparing it to the body of the `runChecks()` function below, it appears that this parameter is meant to indicate whether we simply list some info about the checks (i.e. when `true`) or actually run the checks (i.e. when `false`). 

Why use a parameter to indicate this, instead of a separate function with a more intuitive name?

File: api/router.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		return nil, &httpError{http.StatusInternalServerError, ""}
+	}
+
+	switch checkType {
+	case "node":
+		return rh.runner.PostStart, nil
+	case "cluster":
+		return rh.runner.Cluster, nil
+	}
+	return nil, &httpError{http.StatusNotFound, fmt.Sprintf("unrecognized check type: %s", checkType)}
```
Comment (klueska) [inline api/router.go]: I would probably put a newline above this

File: api/router.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+// http.StatusInternalServerError.
+func (rh *runnerHandler) checkFuncFromReq(r *http.Request) (func(context.Context, bool, ...string) (*runner.CombinedResponse, error), *httpError) {
+	checkType, ok := mux.Vars(r)["check_type"]
+	if !ok {
+		return nil, &httpError{http.StatusInternalServerError, ""}
+	}
+
+	switch checkType {
+	case "node":
+		return rh.runner.PostStart, nil
```
Comment (klueska) [inline api/router.go:98]: I would probably have separate functions for `RunPostStart()`  and `ListPostStart()` rather than `RunPostStart(bool)`, where the boolean indicates run vs. list.

Ditto for below with the Cluster checks.

File: api/router.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+		return checks, nil
+	case "application/json":
+		checks, err := checksFromJSONBody(r)
+		if err != nil {
+			return nil, &httpError{http.StatusBadRequest, err.Error()}
+		}
+		return checks, nil
+	default:
+		return nil, &httpError{http.StatusUnsupportedMediaType, fmt.Sprintf("unsupported Content-Type: %s", ct)}
+	}
```
Comment (klueska) [inline api/router.go]: I'm actually surprised Go is smart enough to recognize that this returns in all cases. I know I've ran into issues with this in the past.

File: api/router.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	err := decoder.Decode(&t)
+	if err != nil {
+		return nil, err
+	}
+
+	checksVal, ok := t["check"]
+	if !ok {
+		return []string{}, nil
+	}
+	checksSlice, ok := checksVal.([]interface{})
```
Comment (klueska) [inline api/router.go]: I would probably have a newline above this

File: api/router.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+	if !ok {
+		return nil, fmt.Errorf("expected checks value of type []string, got %T", checksVal)
+	}
+	checks := make([]string, len(checksSlice))
```
Comment (klueska) [inline api/router.go]: I would probably have a newline above this

File: api/router.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+	for i, check := range checksSlice {
+		check, ok := check.(string)
+		if !ok {
+			return nil, fmt.Errorf("expected check type string, got %T", check)
+		}
+		checks[i] = check
+	}
```
Comment (klueska) [inline api/router.go]: I would probably have a newline above this

File: api/router.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+	err := r.ParseForm()
+	if err != nil {
+		return nil, err
+	}
+
+	checks, ok := r.PostForm["check"]
+	if !ok || checks == nil {
+		return []string{}, nil
+	}
+	return checks, nil
```
Comment (klueska) [inline api/router.go]: I would probably have a newline above this

File: api/router.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// checksFromQueryParams returns a slice of the check names from r's query parameters.
+func checksFromQueryParams(r *http.Request) []string {
+	checks, ok := r.URL.Query()["check"]
+	if !ok || checks == nil {
+		return []string{}
+	}
+	return checks
```
Comment (klueska) [inline api/router.go]: I would probably have a newline above this

File: api/openapi.yaml [hunk 16]
```diff
@@ -0,0 +1,150 @@
+openapi: "3.0.1"
```
Comment (klueska) [inline api/openapi.yaml:1]: When / how is this file used?

File: README.md [hunk 17]
```diff
@@ -23,13 +23,13 @@ make test
   dcos-check-runner check <check-type> [flags]
```
Comment (klueska) [inline README.md:23]: There doesn't seem to be any help info on the valid `<check-type>` values to pass in.

File: cmd/root.go [hunk 18]
```diff
@@ -59,6 +64,12 @@ func Execute() {
 func init() {
 	cobra.OnInitialize(initConfig)
 
+	if runtime.GOOS == "windows" {
+		defaultCheckConfig = os.Getenv("SYSTEMDRIVE") + defaultCheckConfigWindows
+	}
+
+	RootCmd.PersistentFlags().StringVar(&checkCfgFile, "check-config", defaultCheckConfig,
+		"Path to check configuration file")
```
Comment (klueska) [inline cmd/root.go:72]: I would have written this as either:
```
RootCmd.PersistentFlags().StringVar(
    &checkCfgFile,
    "check-config",
    defaultCheckConfig,
    "Path to check configuration file")
```
or
```
RootCmd.PersistentFlags().StringVar(
    &checkCfgFile, "check-config", defaultCheckConfig, "Path to check configuration file")
```

The logic being that if a list of parameters can't fit nicely onto a single line, then each parameter should go on its own line.

If this is always done consistently, it makes things very clear without having to bend your mind around looking at how many parameters are being passed to a function and what they are.

File: README.md [hunk 19]
```diff
@@ -23,13 +23,13 @@ make test
...
-      --list                  List runner
+  -h, --help   help for check
+      --list   List runner
 
 Global Flags:
-      --config string   config file (default is /opt/mesosphere/etc/dcos-check-runner.yaml)
-      --role string     Set node role
-      --verbose         Use verbose debug output.
-      --version         Print dcos-check-runner version
+      --check-config string   Path to check configuration file (default "/opt/mesosphere/etc/dcos-check-config.json")
```
Comment (klueska) [inline README.md:30]: What is the difference between the `check-config` and the "normal" `config`?

File: README.md [hunk 20]
```diff
@@ -33,3 +33,21 @@ Global Flags:
       --verbose               Use verbose debug output.
       --version               Print dcos-check-runner version
 ```
+
+```
+  dcos-check-runner http-server [flags]
```
Comment (klueska) [inline README.md]: I'm not sure how I feel about the name `http-server` as the subcommand here. I guess it's a bit of an internal detail as we only run this binary inside DC/OS, but running a command for `dcos-check-runner http-server` implies to me that an HTTP server is going to somehow get spawned off by this process, and then exit -- not that this process will *become* the HTTP server and then continue long-running.

File: cmd/http_server.go [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+		router := api.NewRouter(r, defaultConfig.FlagBaseURI)
+		if defaultConfig.FlagSystemdSocket {
+			listener, err := getSystemdSocket()
+			if err != nil {
+				logrus.Fatal(err)
+			}
+			logrus.Infof("Listening at %s", listener.Addr().String())
+			logrus.Fatal(http.Serve(listener, router))
+		} else {
```
Comment (klueska) [inline cmd/http_server.go]: Do you need the `else` here? It looks like you always fatally exit above if listening fails.

File: api/logging.go [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+			"user_agent":  r.UserAgent(),
+			"method":      r.Method,
+			"uri":         r.URL.RequestURI(),
+		}))
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// logRequestsMiddleware returns a http.Handler that writes a log entry for each request and response.
+func logRequestResponseMiddleware(next http.Handler) http.Handler {
```
Comment (klueska) [inline api/logging.go]: The comment here doesn't match the function name

File: api/router.go [hunk 23]
```diff
@@ -17,12 +18,24 @@ func NewRouter(runner *runner.Runner, baseURI string) *mux.Router {
...
 	base := router.PathPrefix(baseURI).Subrouter()
-	base.HandleFunc("/{check_type}/", rh.listChecks).Methods("GET")
-	base.HandleFunc("/{check_type}/", rh.runChecks).Methods("POST")
+	base.Handle("/{check_type}/", withMiddlewares(http.HandlerFunc(rh.listChecks))).Methods("GET")
+	base.Handle("/{check_type}/", withMiddlewares(http.HandlerFunc(rh.runChecks))).Methods("POST")
 
 	return router
 }
 
+var middlewares = [...]func(http.Handler) http.Handler{
```
Comment (klueska) [inline api/router.go]: Why does this need to be global?

File: api/router.go (hunk 2 above)
Comment (branden) [reply api/router.go]: I think you left this comment on an older version of the PR, which is why it shows up as "outdated". This PR currently declares `runnerHandler` on line 39 of `api/router.go`.

I have `NewRouter()` at the top because you recommended that I put exported stuff at the top of the file. Should I put type declarations at the top ahead of exported names? I think it makes sense to define types and their methods next to each other, so that'd mean moving `NewRouter()` down even further in the file.

File: api/router.go (hunk 4 above)
Comment (branden) [reply api/router.go]: The best I can come up with is to use `checker` instead of `checkFunc`, which sounds grating to my ears. I'll just put a `get` in front to remove the ambiguity.

File: api/router.go (hunk 5 above)
Comment (branden) [reply api/router.go:207]: See https://github.com/dcos/dcos-check-runner/pull/7#discussion_r203915930

File: api/router.go (hunk 7 above)
Comment (branden) [reply api/router.go:50]: Agreed, the semantics changing so much based on the value of a bool param suggests that listing and running checks should be separate methods. But that refactor wasn't necessary to implement the API, so I didn't bother with it. And before I do that refactor I want to add tests for the runner CLI. Most of the existing tests in this repo are just unit tests against `CheckRunner` methods.

File: api/router.go (hunk 9 above)
Comment (branden) [reply api/router.go:98]: See https://github.com/dcos/dcos-check-runner/pull/7#discussion_r203916028

File: api/router.go (hunk 10 above)
Comment (branden) [reply api/router.go]: :D I'll remove the `default` case and put this line after the `switch`.

File: api/openapi.yaml (hunk 16 above)
Comment (branden) [reply api/openapi.yaml:1]: Nothing uses it currently, but there are a bunch of doc generators that read this format. I added it here as a courtesy to anyone working on the API. I'll eventually copy it to https://github.com/mesosphere/dcos-docs-site, where we'll generate a nice-looking API doc from it.

File: README.md (hunk 17 above)
Comment (branden) [reply README.md:23]: That's a problem, but I don't think fixing it is in the scope of this PR.

File: README.md (hunk 19 above)
Comment (branden) [reply README.md:30]: `check-config` (not a good name) contains check definitions; `config` is actual config for the CLI or API (role, port, etc.)

File: cmd/http_server.go (hunk 21 above)
Comment (branden) [reply cmd/http_server.go]: Don't need it, but it seems clearer to have the two different listening methods (systemd socket vs. port) under the `if`.  We're choosing between two alternatives, not conditionally executing a single stanza. Looking at this more, I think I actually want to end these clauses by assigning to an `error` and then call `logrus.Fatal()` after the `if`/`else`, since that'll more clearly express that we always do that.

File: api/router.go (hunk 23 above)
Comment (branden) [reply api/router.go]: It doesn't, I'll move it into `withMiddlewares()`.

File: README.md (hunk 20 above)
Comment (branden) [reply README.md]: Do you have an example of something behaving that way? `http-server` doesn't imply "fork a server and exit" to me (`daemon` sounds more appropriate for that). `python -m http.server` comes to mind.

File: cmd/root.go (hunk 18 above)
Comment (branden) [reply cmd/root.go:72]: I wanted to stay consistent with the style currently used in this file. I'd like to wait until I have a PR that does more work here before I overhaul its style, or maybe do that as its own PR.

**Comment by branden** [comment]
@klueska Comments addressed, PTAL. I think this is all wrapped up except for https://github.com/dcos/dcos-check-runner/pull/7#discussion_r203915930, where I'll need your response.

File: api/router.go (hunk 2 above)
Comment (klueska) [reply api/router.go]: I just go through commit-by-commit adding comments. If I try and look at the PR as a whole, I find it hard to follow. So I probably left this comment on the first commit and then you changed something around it in a later commit, making it look outdated.

I actually think the way you have it is fine.

I would probably done it as follows, but I don't think it's worth changing:
```
<type definitions>
<exported functions>
<type-specific functions>
<package-local functions>
```

**Review by klueska — APPROVED** [review]
